crates/circuit/src/symbol_expr.rs

```diff
@@ -0,0 +1,1528 @@
+    Value(Value),
+    Unary(Arc<Unary>),
+    Binary(Arc<Binary>),
+}
+
+#[derive(Debug, Clone)]
+pub struct Symbol {
+    name : String,
+}
+
+// ================================
+// real number and complex number
+// (separate for performance)
+// ================================
+#[derive(Debug, Clone)]
+pub enum Value {
+    Real(f64),
+    Complex(Complex64),
+}
+
+// ================================
+// Operators
+// ================================
+#[derive(Debug, Clone, PartialEq)]
+pub enum UnaryOps {
+    Abs,
+    Neg,
+    Sin,
+    Asin,
+    Cos,
+    Acos,
+    Tan,
+    Atan,
+    Exp,
+    Log,
+}
+
+#[derive(Debug, Clone, PartialEq)]
+pub enum BinaryOps {
+    Add,
+    Sub,
+    Mul,
+    Div,
+    Pow,
+}
+
+#[derive(Debug)]
+pub struct Unary {
+    op : UnaryOps,
+    expr : SymbolExpr,
+}
+
+#[derive(Debug)]
+pub struct Binary {
+    op : BinaryOps,
+    lhs : SymbolExpr,
+    rhs : SymbolExpr,
+}
+
+impl Clone for SymbolExpr {
+    fn clone(&self) -> Self {
+        match self {
+            SymbolExpr::Symbol(e) => SymbolExpr::Symbol(e.clone()),
+            SymbolExpr::Value(e) => SymbolExpr::Value(e.clone()),
+            SymbolExpr::Unary(e) => SymbolExpr::Unary(e.clone()),
+            SymbolExpr::Binary(e) => SymbolExpr::Binary(e.clone()),
+        }
+    }
+}
+
+impl SymbolExpr {
+    pub fn to_string(&self) -> String {
+        match self {
+            SymbolExpr::Symbol(e) => e.to_string(),
+            SymbolExpr::Value(e) => e.to_string(),
+            SymbolExpr::Unary(e) => e.to_string(),
+            SymbolExpr::Binary(e) => e.to_string(),
+        }
+    }
+
+    pub fn subs(&self, maps: &HashMap<String, SymbolExpr>) -> SymbolExpr {
+        match self {
+            SymbolExpr::Symbol(e) => e.subs(maps),
+            SymbolExpr::Value(e) => SymbolExpr::Value(e.clone()),
+            SymbolExpr::Unary(e) => e.subs(maps),
+            SymbolExpr::Binary(e) => e.subs(maps),
+        }
+    }
+
+    pub fn eval(&self, recurse: bool) -> Option<Value> {
+        match self {
+            SymbolExpr::Symbol(_) => None,
+            SymbolExpr::Value(e) => Some(e.clone()),
+            SymbolExpr::Unary(e) => e.eval(recurse),
+            SymbolExpr::Binary(e) => e.eval(recurse),
+        }
+    }
+
+    pub fn derivative(&self, param: &String) -> SymbolExpr {
+        match self {
+            SymbolExpr::Symbol(e) => if e.name == *param {
+                SymbolExpr::Value( Value::Real(1.0))
+            } else {
+                SymbolExpr::Value( Value::Real(0.0))
+            },
+            SymbolExpr::Value(e) => SymbolExpr::Value( Value::Real(0.0)),
+            SymbolExpr::Unary(e) => e.derivative(param),
+            SymbolExpr::Binary(e) => e.derivative(param),
+        }
+    }
+
+    pub fn real(&self) -> f64 {
+        match self.eval(true) {
+            Some(v) => match v {
+                Value::Real(r) => r,
+                Value::Complex(c) => c.re,
+            }
+            None => 0.0,

```

Comment:
I might not understand this fully, but if `self` is a plain `Symbol` then `self.eval` returns `None` and this the functions `real/complex` will return 0, right? But if I just have a plain `Symbol` with no value associated yet, then I would expect that I cannot query the real or complex parts, because they are not defined, no? (In which case I thought this might need to raise an error 🙂)